### ISSUE TITLE
Issue when aborting a transaction in progress

### ISSUE THREAD

**Issue by Jamie5** [berniegp/mock-xmlhttprequest#30]
**Is this issue a *feature* request or a *bug* report?**

Bug report

**What is the current behavior?**

1. `xhr.send` gets called
2. In `send`, the following happens: `setTimeout(() => onSend.call(this, this), 0)`
3. Meanwhile, the code will call `xhr.abort()`
4. Now, `onSend` gets called and by default goes to `_handleRequest`, which will eventually call request handlers even though the request was already aborted. The request handlers might e.g. check the `url`, which has been deleted by `_terminateRequest`.

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem.**

**What is the expected behavior?**

Since the request was already aborted, we shouldn't be hitting the handlers or anything.

**Environment in which you encounter the issue:**

- mock-xmlhttprequest version: 7.0.4
- Node.js version: N/A
- OS version: 

**Did this work in a previous version of mock-xmlhttprequest (which)?**

**Do you intend to provide a pull request to address this issue?**

**Comment by berniegp** [comment]
Thanks for the report. Does this test correctly match your issue?

```
const xhr = new MockXhr();

// Add onSend callback
let onSendCount = 0;
xhr.onSend = () => {
  onSendCount += 1;
};

// Aborted send() during the loadstart event handler
xhr.open('GET', '/url');
xhr.send();
xhr.abort();

setTimeout(() => {
  assert.strictEqual(onSendCount, 0, 'onSend() should not be called');
}, 100);
```

**Comment by Jamie5** [comment]
Yeah, though I guess in our case it is due to `(mockServer: MockXhrServer).setDefaultHandler(handler)` and it is the handler which looks at the request URL and stuff (and hence fails when the URL is not present)

**Comment by berniegp** [comment]
I see. We're hitting a fine point about what the expected behavior should be. I _think_ `send()` as written in the spec means that a real server _can_ have received the request even if you `abort()` right after. In that case, I think it makes sense for the `MockXhrServer` to handle the (eventually aborted) request.

But I do see that the wrong route will be chosen (i.e. the default handler) because `MockXhr.url` will have been deleted. The server's request log will also have "bad" data. I think those 2 points are really where the bug lies.

What do you think? I also understand that for your use-case you'd rather not have the request reach the server at all if it's aborted before the server has the time to call it's handler.

As a (maybe temporary) suggestion, perhaps your handler can check that `xhr.readyState !== MockXhr.UNSENT`?

**Comment by Jamie5** [comment]
Hmm you make a good point regarding that it could still hit the server. In our case we actually only use the default handler (there are reasons that make doing our own thing for sorting out URLs useful for various reasons), and are working around things by checking if the url is undefined. So I guess for the common case that you point out where people would just register URLs and have handlers for them, and the logs, is the issue, and I guess fixing those would be good things to do.

**Comment by berniegp** [comment]
There's another more involved "fix" where I can add a *synchronous* version of `onSend` for the server to:
1. Register the request with the original URL (or even make a copy of the xhr's attribute values?)
2. Schedule a handler callback
3. Provide the "original URL and method" information to the handler along with the xhr so it can decide what to do

The more I think about it though, the more it seems like we need a more robust system to make sure mock server responses match the actual requests. Consider these events:
1. `send()` to `/url1`
2. `abort()`
3. `open()` + `send()` to `/url2`
In this scenario, even if a response is received by the client for the `/url1` request, it will be discarded in favor of the response to `/url2`.

In other words, the server's view of receiving an HTTP request should be independent from whether that HTTP request is still active on the client.

**Comment by berniegp** [comment]
To address what you said about using only the default handler, it's not a use-case I thought about (i.e. externalizing request handling) when combined with `MockXhrServer` but I can see that it makes sense.

Although I'm curious to know what you gain with using `MockXhrServer` instead of `newMockXhr()` directly.

**Comment by Jamie5** [comment]
Are you actually allowed to call send twice on the same request? I believe MockXhr throws in that case and https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/send also claims that calling send twice results in an exception.

As for why not just newMockXhr directly, I guess the quick start seemed reasonable enough so went with that and didn't read on. And I guess it does say it's mostly noted for historical reasons.

**Comment by berniegp** [comment]
`send()` twice isn't allowed. However you can string together a bunch of `open()` and `send()` sequences and drop some `abort()`s in there if you like.

Some sources:
https://xhr.spec.whatwg.org/#the-open()-method
https://github.com/whatwg/xhr/issues/88

> As for why not just newMockXhr directly, I guess the quick start seemed reasonable enough so went with that and didn't read on. And I guess it does say it's mostly noted for historical reasons.
Fair enough!

**Comment by berniegp** [comment]
There are a lot of tests here if you want to do a deep dive: https://github.com/web-platform-tests/wpt/tree/master/xhr

**Comment by berniegp** [comment]
@Jamie5 would love to hear if the merged fix fixes your issue.

**Comment by Jamie5** [comment]
Ran into fun webpack issues while trying to see, but based on the code it looks very promising! Thank you!

**Comment by berniegp** [comment]
This is fixed in v8.0.0.
